heapster-saw/doc/tutorial/tutorial.md

```diff
@@ -0,0 +1,623 @@
+[16:49:13.647] Loading module ffs.bc
+[16:49:13.651] Extracting reference term: ffs_ref
+[16:49:13.663] Extracting implementation term: ffs_imp
+[16:49:13.666] Proving equivalence: ffs_ref == ffs_imp
+[16:49:13.716] Valid
+[16:49:13.716] Done.
+```
+
+That's it! You know the basics of SAW.
+
+## Using Heapster
+
+Heapster is, fundamentally, a type system for extracting functional
+specifications from memory-safe imperative programs. The type system,
+defined inside SAW, uses separation types to reason about memory
+safety. Once a program is type-checked as memory-safe, it can be then
+extracted as a functional program to be verified in Coq.
+
+**TODO: Double check this description of Heapster**
+
+### Heapster type-checking overview
+
+Heapster allows us to (1) type check programs with respect to
+types/specifications that can express separation loigc and (2) extract
+the resulting functional program to Coq for further verification.
+
+The process will generally involve
+
+- [1. Generating LLVM bitcode](#1-generating-llvm-bitcode)
+- [2. Run the SAW interpreter with Heapster](#2-run-the-saw-interpreter-with-heapster)
+- [3. Load the file and extract the two function specifications.](#3-load-the-file-and-extract-the-two-function-specifications-1)
+- [4. Writing heapster specifications for your functions](#4-writing-heapster-specifications-for-your-functions)
+- [5. Writing a SAW script to type-check your code with respect to the sepcification](#5-writing-a-saw-script-to-type-check-your-code-with-respect-to-the-sepcification)
+- [6. Writing a Coq file to prove things about the generated functional specification(s)](#6-writing-a-coq-file-to-prove-things-about-the-generated-functional-specifications)
+
+Just like with SAW, Heapster can be processed in batch. To do so, you
+can combine steps 2-6 in a `.saw` file and use SAW's batch processing.
+
+### Running an example
+
+This section will walk through the process of using Heapster to write
+and verify some C code. Specifially, we want to verify the function
+`is_elem`, which tests if a specific value is in a list. The function,
+together with others, can be found in `examples/linked_list.c`.
+
+```C
+typedef struct list64_t {
+  int64_t data;
+  struct list64_t *next;
+} list64_t;
+
+int64_t is_elem (int64_t x, list64_t *l) {
+  if (l == NULL) {
+    return 0;
+  } else if (l->data == x) {
+    return 1;
+  } else {
+    return is_elem (x, l->next);
+  }
+}
+```
+
+#### 1. Generating LLVM bitcode
+
+Just like with SAW, we want to work with the LLVM bitcode (`.bc`), so
+you can generate it just like before. Note that we have already
+included the binary for all the examples so you don't really need to
+run this command.
+
+```bash
+clang -g -c -emit-llvm -o examples/linked_list.bc examples/linked_list.c
+```
+
+(The alternative command `make examples/linked_list.bc` won't work
+because the shorcut is not defined in the Makefile of heapster.)
```

Comment:
I'm confused by this comment, because `make linked_list.bc` works for me (when run in the `heapster-saw/examples` directory)